pkg/auth/dependency/interaction/adaptors/authenticator.go

```diff
@@ -157,6 +157,52 @@ func (a *AuthenticatorAdaptor) List(userID string, typ interaction.Authenticator
 	return ais, nil
 }
 
+func (a *AuthenticatorAdaptor) ListByIdentity(userID string, ii *interaction.IdentityInfo) (ais []*interaction.AuthenticatorInfo, err error) {
+	// This function takes IdentityInfo instead of IdentitySpec because
+	// The login ID value in IdentityInfo is normalized.
+	switch ii.Type {
+	case authn.IdentityTypeOAuth:
+		// OAuth Identity does not have associated authenticators.
+		return
+	case authn.IdentityTypeLoginID:
+		// Login ID Identity has password, TOTP and OOB OTP.
+		// Note that we only return OOB OTP associated with the login ID.
+		var pas []*password.Authenticator
+		pas, err = a.Password.List(userID)
+		if err != nil {
+			return
+		}
+		for _, pa := range pas {
+			ais = append(ais, passwordToAuthenticatorInfo(pa))
+		}
+
+		var tas []*totp.Authenticator
+		tas, err = a.TOTP.List(userID)
+		if err != nil {
+			return
+		}
+		for _, ta := range tas {
+			ais = append(ais, totpToAuthenticatorInfo(ta))
+		}
+
+		loginID := ii.Claims[interaction.IdentityClaimLoginIDValue]
+		var oas []*oob.Authenticator
+		oas, err = a.OOBOTP.List(userID)
+		if err != nil {
+			return
+		}
+		for _, oa := range oas {
+			if oa.Email == loginID || oa.Phone == loginID {
```

Comment:
I suppose we don't need to check the type of login ID key.

Reply:
This if statement makes sure we only include OOB Authenticators that is related to the given identity.